Add base for VSTS web hook

olivierdagenais opened the pull request:
These changes add a framework for a VSTS endpoint, with the events we expect to be able to receive from Visual Studio Team Services and Team Foundation Server.

The `VstsWebHook` exposes the `/vsts-hook/` URL where payloads may be POSTed with an event name.

Manual testing
--------------

1. Run `mvn hpi:hpl` to prepare the `tfs.hpl` file, then fix it up to add the TFS SDK JAR to the `CLASSPATH`.  This will allow us to test our plugin in a local Jenkins instance without having to re-deploy.
2. Launch the local Jenkins instance on port 9090: `java -jar jenkins.war --httpPort=9090`
    1. You can see it is responding at `http://localhost:9090`
3. Use cURL to POST to the new `/vsts-hook/` endpoint in `PING` mode, sourcing the body of the request from the `git_code_pushed.json` text file: `curl --request POST http://localhost:9090/vsts-hook/?event=PING --header 'Content-Type: application/json' --upload-file git_code_pushed.json`
    1. In the console, we see the following output:

    ```
    Pong from the Jenkins TFS plugin! Here's your body:
    {
        "repoUrl":"https://mseng.visualstudio.com/Personal/_git/olivida.gcm4ml",
        "serverBaseUrl":"https://mseng.visualstudio.com",
        "projectId":"Personal",
        "repoId":"olivida.gcm4ml",
        "commit":"6a23fc7afec31f0a14bade6544bed4f16492e6d2",
        "pushedBy":"olivida"
    }
    ```
    ...which just so happens to be the contents of the `git_code_pushed.json` file.

Mission accomplished!

DavidStaheli commented on src/main/java/hudson/plugins/tfs/VstsHookEventName.java:
```diff
@@ -0,0 +1,41 @@
+    PULL_REQUEST_MERGE_COMMIT_CREATED,
+    // TODO: clarify the following events
+    DEPLOYMENT_COMPLETED,
+    RELEASE_CREATED,
```
From feedback, I learned that the `Release created` event is not something we want to respond to.  It occurs when a VSTS release is instigated, but not yet completed, and perhaps not yet even approved to begin.  On the other hand, the `Deployment completed` event indicates that one of the release's deployments have completed successfully or not.

olivierdagenais replied:
Documented `DEPLOYMENT_COMPLETED` in a761b0b and removed `RELEASE_CREATED` in 22d08b5.

DavidStaheli commented on src/main/java/hudson/plugins/tfs/VstsWebHook.java:
```diff
@@ -0,0 +1,101 @@
+        List<? extends GitStatus.ResponseContributor> contributors = null;
+        switch (event) {
+            case PING:
+                final String message = "Pong from the Jenkins TFS plugin! Here's your body:\n" + parsedBody;
```
P3: "Jenkins TFS plugin" -> "TFS Plugin for Jenkins" ?

olivierdagenais replied:
Good question!  I've been calling it "Jenkins TFS plugin" for what seems like forever, yet there appears to be inconsistent usage.  I'll revisit this later.

DavidStaheli commented on src/main/java/hudson/plugins/tfs/VstsWebHook.java:
```diff
@@ -0,0 +1,101 @@
+                public void generateResponse(StaplerRequest req, StaplerResponse rsp, Object node)
+                        throws IOException, ServletException {
+                    rsp.setStatus(SC_OK);
+                    rsp.setContentType("text/plain");
```
Is `javax.ws.rs.core.MediaType.TEXT_PLAIN` available to use here?

DavidStaheli commented on src/main/java/hudson/plugins/tfs/util/StringBodyParameter.java:
```diff
@@ -0,0 +1,54 @@
+
+            final String contentType = request.getContentType();
+
+            if ("application/json".equals(contentType)) {
```
Is `javax.ws.rs.core.MediaType.APPLICATION_JSON_TYPE` available to use here?

DavidStaheli commented on src/main/java/hudson/plugins/tfs/util/StringBodyParameter.java:
```diff
@@ -0,0 +1,54 @@
+                    LOGGER.log(Level.SEVERE, "Unable to obtain request body: {}", e.getMessage());
+                }
+            }
+            else if ("application/x-www-form-urlencoded".equals(contentType)) {
```
Is `javax.ws.rs.core.MediaType.APPLICATION_FORM_URLENCODED` available to use here?

DavidStaheli commented:
Looks good!  Please see minor comments.

olivierdagenais commented:
`javax.ws.rs.core.MediaType` wasn't available, so I added `hudson.plugins.tfs.util.MediaType` in 0cb222a3b2a0c99c253025b6905402ee32fb58de.

I also removed duplicate calls to `Charset.forName` and extracted a constant in `MediaType`.  This made me wonder about the use of character encodings, so in f6222e372bb0a1e00337195b6c550e8fe33f1fad I made a few improvements that enabled me to re-run the PING test with cURL that included a body that was `UTF-8` encoded.  Without that last change, the response encoding defaulted to `ISO-8859-1` and the response body was corrupted.

olivierdagenais commented:
Thank you for your code review!
